pkg/apiserver/rest/dualwriter.go

```diff
@@ -217,6 +232,26 @@ func SetDualWritingMode(
 		}
 	}
 
+	if (desiredMode == Mode3) && (currentMode == Mode2) {
+		// This is where we go through the different gates to allow the instance to migrate from mode 2 to mode 3.
+
+		// gate #1: ensure the data is 100% in sync
+		syncOk, err := runDataSyncer(ctx, currentMode, legacy, storage, entity, reg, serverLockService, requestInfo)
+		if err != nil {
+			klog.Info("data syncer failed for mode:", m)
+		} else {
+			if syncOk {
+				// only set currentMode to 3 if data is in sync
+				currentMode = Mode3
+			}
+		}
+
+		err = kvs.Set(ctx, entity, fmt.Sprint(currentMode))
+		if err != nil {
+			return Mode0, errDualWriterSetCurrentMode
```

Comment:
with the changes above we could use `desiredMode` on line 249 instead of overwriting `currentMode`, which would allow us to return `currentMode` here instead of `Mode0`.

Reply:
Makes sense. I modified the code to reflect that. We will still need to refactor the other transition blocks, but I will address those in a follow-up PR